msbuild Restore and Visual Studio produce different assets when assembly name if different from project name

erdembayar opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/8272

Regression? Last working version: N/A
Current issue only applies to legacy packageReference packages, not sdk style.
msbuild /t:Restore and Visual Studio restore produce different assets files when assembly name <> project name.
When you perform an msbuild /restore the project.assets.json that gets generated uses the AssemblyName of the referenced class Library. However if you open up Visual Studio it will perform a restore setting this to the name of the referenced PROJECT file. This change forces a Visual Studio rebuild. 
Here is fix to for to make VS and MSbuild generate same asset file.
![image](https://user-images.githubusercontent.com/8766776/127401466-84399f58-2416-4683-8a26-33eaacf80d61.png)

## Description

## PR Checklist

- [x] PR has a meaningful title
- [ ] PR has a linked issue.
- [ ] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

erdembayar committed:
AssemblyName is considered during legacy packagereference project asset file creation.

erdembayar committed:
Remove unnecessary check

erdembayar commented:
@NuGet/nuget-client 
Please review.

nkolev92 reviewed (commented):
The project name tends to be: 

PackageId -> AssemblyName -> Project File Name. 

Can you check if setting a PackageId in a legacy PR project suffers from the same issue?

nkolev92 commented:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L672-L684 has a comment which is probably incorrect, when it says non-netcore it really means non PackageReference.

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.

erdembayar committed:
Address latest PR comment

erdembayar committed:
Clean up.

erdembayar commented:
> The project name tends to be:
> 
> PackageId -> AssemblyName -> Project File Name.
> 
> Can you check if setting a PackageId in a legacy PR project suffers from the same issue?

Thank you for your review. 
Yes, it suffers from same problem. I added PackageId in latest commit. Please check I added End2end test too.
Please review again.

erdembayar commented:
@nkolev92 
Please review when you have time.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:
```diff
@@ -129,6 +128,16 @@ private async Task<bool> IsCentralPackageManagementVersionsEnabledAsync()
             return MSBuildStringUtility.IsTrue(await _vsProjectAdapter.GetPropertyValueAsync(ProjectBuildProperties.ManagePackageVersionsCentrally));
         }
 
+        private async Task<string> GetSpecifiedAssemblyName()
```
Probably want to add async to the name.

erdembayar replied:
Merged without addressing this. I'll make make another PR later.
